docs/design/specs/Ecma-335-Augments.md

```diff
@@ -420,3 +420,393 @@ The algorithm is amended as follows:
+static or instance members) or overrides (for virtual instance methods) a member from a base class or interface,
+simply substitute each generic parameter with its generic argument, and compare the resulting member
+signatures. [*Example*: The following illustrates this point:
+
+### II.9.11, Constrains on Generic Parameters
+
+(Change first paragraph)
+
+A generic parameter declared on a generic class or generic method can be *constrained* by one or more
+type (for encoding, see *GenericParamConstraint* table in paragraph II.22.21) and by one or more special
+constraints (paragraph II.10.1.7). Generic parameters can be instantiated only with generic arguments that are
+*assignable-to* (paragraph I.8.7.3) (when boxed) and *implements-all-static-interface-methods-of* (**paragraph
+reference needed**) each of the declared constraints and that satisfy all specified special constraints.
+
+(Change the last paragraph on page 137)
+
+[*Note*: Constraints on a generic parameter only restrict the types that the generic parameter may
+be instantiated with. Verification (see Partition III) requires that a field, property or method that a
+generic parameter is known to provide through meeting a constraint, cannot be directly
+accessed/called via the generic parameter unless it is first boxed (see Partition III) or the **callvirt**,
+**call** or **ldftn** instruction is prefixed with the **constrained.** prefix instruction (see Partition III). *end note*]
+
+### II.10.3 Introducing and overriding virtual methods
+
+(Change first paragraph)
+
+A virtual method of a base type is overridden by providing a direct implementation of the method
+(using a method definition, see paragraph II.15.4) and not specifying it to be newslot (paragraph II.15.4.2.3). An existing
+method body can also be used to implement a given instance or static virtual declaration using the .override directive
+(paragraph II.10.3.2).
+
+### II.10.3.2 The .override directive
+
+(Change first paragraph)
+
+The .override directive specifies that a virtual method shall be implemented (overridden), in this type,
+by a virtual instance method with a different name or a non-virtual static method, but with the same signature.
+This directive can be used to provide an implementation for a virtual method inherited from a base class, or
+a virtual method specified in an interface implemented by this type. The .override directive specifies a Method
+Implementation (MethodImpl) in the metadata (§II.15.1.4).
+
+(Change the third and fourth paragraph on page 148, the second and third one below the table)
+
+The first *TypeSpec::MethodName* pair specifies the virtual method that is being overridden, and shall
+be either an inherited virtual method or a virtual method on an interface that the current type
+implements. The remaining information specifies the virtual instance or non-virtual static method that
```

Comment:
Should we explicitly disallow mismatch of instance vs. static in the pair?

Reply:
The existing ECMA spec says in the .override section (II.10.3.2, page 147):

*The .override directive specifies that a virtual method shall be implemented (overridden), in this type,
by a virtual method with a different name, but with the same signature. This directive can be used to
provide an implementation for a virtual method inherited from a base class, or a virtual method
specified in an interface implemented by this type.*

As you see, static may be used to override an instance method (there's no harm in dropping the this pointer when not needed), of course not the other way round, that may be worth clarifying.

Reply:
Sorry, I looked at the updated comment, the original ECMA spec doesn't allow overriding virtuals with statics, I need to double-check how it got to the update - I think it was meant to clarify the distinction between virtual instance and virtual static methods but I think it's quite confusing the way it stands now, I'll fix that.

Reply:
Updated in 12th commit - @davidwrighton, can you please take a look at the wording change I made in that commit and let me know if you find that reasonable?

Reply:
@AlekseyTs, we already explicitly disallow mismatch of instance vs static as the signature in Ecma335 includes the detail of whether or not a method is instance vs static. I don't like the additional change @trylek added in change 12 which is redundant. I would prefer to revert back to the wording in change 11.

Reply:
@davidwrighton 

>  we already explicitly disallow mismatch of instance vs static as the signature in Ecma335 includes the detail of whether or not a method is instance vs static.

Could you point me to the relevant section of provide a quote from it.

Reply:
I can easily revert the 12th commit but in light of your clarifications I'm wondering what was the purpose of adding the mention of non-virtual statics to the ECMA spec in the first place?

*[DavidWr] In section 10.3.2, replace ... "by a virtual method with a different name" with "by a virtual instance method with a different name, **or a non-virtual static method**"*

Reply:
@AlekseyTs The definition of a method signature is in I.8.6.1.5, and the quote is "A method signature is composed of 
 a calling convention, 
 the number of generic parameters, if the method is generic, "

Also, II.15.3 where there is a definition of calling convention which describes how to specify an instance method (as opposed to a static method.)

The spec is somewhat inconsistent in defining whether or not a calling convention includes the static vs instance detail as in partition I.8.6.1.6 it calls out calling conventions and static vs instance separately. But even there, a matching signature is considered to include details such as if it is instance vs static.

@trylek The point of my wording is that we are not intending to allow virtual static methods to exist on classes. Thus any implementation itself cannot be virtual. The change in wording is to allow a non-virtual static method to implement a virtual static interface method. This mixing of virtual and non-virtual is not permitted for interface methods (even though C# allows such logic. C# implements such methods by silently making those methods virtual).